Add placeholder for job managed-by label docs [placeholder] (#45132)
kubernetes/website: content/en/docs/concepts/workloads/controllers/job.md

  terminating: 3 # three Pods are terminating and have not yet reached the Failed phase
```

### Delegation of managing a Job object to external controller

> {{< feature-state for_k8s_version="v1.30" state="alpha" >}}

Reviewer: We should mention that feature gate should be enabled to use this feature.

Can we also mention the risks of disabling the feature gate or downgrading to pre-alpha version?

Author: Sure, I will update it.

Author: Updated, PTAL

This feature allows you to disable the built-in Job controller, for a specific
Job, and delegate reconciliation of the Job to an external controller.

By setting a custom value of the `spec.managedBy` field - any value
other than `kubernetes.io/job-controller` - you indicate the controller which
reconciles the Job. The value of the field is immutable.

{{< note >}}
When using this feature make sure the controller indicated by the field is
installed, otherwise the Job may not be reconciled at all.
{{< /note >}}

{{< note >}}

> When developing an external Job controller make sure it does not use the

Reviewer: Should we mention that the controllers should conform to the API types, API comments? Not sure if we can define the expected Job behaviour in other way...

And also pass the conformance tests? Or other tests as well?

Author: We could say certainly say that the controller status updates should pass the validation rules and be conformant with the API field definitions in the API comments. 

However, running the conformance tests as such might be infeasible. For example in case of MultiKueue multiple conformance tests also assert on pods, which would not pass for the controller, for example here: https://github.com/kubernetes/kubernetes/blob/246e678acc97d68261ce3e9ae858332f01b44058/test/e2e/apps/job.go#L94-L98. Also, I don't think we will have any power to enforce some external controller was validated against the conformance tests. I think the status validation is the tool to enforce that conformance in fact.

 I will update the note.

Reviewer: I think, we should say that they should also be conformant with the spec.

I see. Nevertheless, we could write about the compliance with the conformance tests in more generic way. That is if you run the pods somewhere else, you could consider that and update the places in the tests. I suppose you might not even run the pod/executable in kubernetes.

Yes, we do not have any power to enforce this, but if we get a bug that a job is behaving in an unsupported way, we can point to the documentation and say that the external controller should comply.

@alculquicondor @soltysh for additional opinions

Author: Updated with some initial formulation. PTAL.

Reviewer: lgtm, thanks for updating the section

> `batch.kubernetes.io/job-tracking` finalizer, reserved for the built-in controller.

Reviewer: Just want to confirm. This will not apply for job mirroring, right?

Author: Correct, actually for MultiKueue we only mirror the status from worker to management cluster. There are no pods created by the Job on the management cluster.

  terminating: 3 # three Pods are terminating and have not yet reached the Failed phase
```

### Delegation of managing a Job object to external controller

{{< feature-state for_k8s_version="v1.30" state="alpha" >}}

{{< note >}}

> You can only set `ManagedBy` field on Jobs if you enable the `JobManagedBy`

Reviewer 2: ```suggestion
You can only set the `managedBy` field on Jobs if you enable the `JobManagedBy`
```

Author: done

[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
(disabled by default).
{{< /note >}}

This feature allows you to disable the built-in Job controller, for a specific
Job, and delegate reconciliation of the Job to an external controller.

> By setting a custom value of the `spec.managedBy` field - any value

Reviewer 2: You indicate the controller that reconciles the Job by setting a custom value for....

Author: Done

other than `kubernetes.io/job-controller` - you indicate the controller which
reconciles the Job. The value of the field is immutable.

{{< note >}}

> When using this feature make sure the controller indicated by the field is

Reviewer 2: ```suggestion
When using this feature, make sure the controller indicated by the field is
```

Author: Done

installed, otherwise the Job may not be reconciled at all.
{{< /note >}}

{{< note >}}
When developing an external Job controller be aware that your controller needs
to operate in a fashion conformant with the definitions of the API spec and
status fields of the Job object.

> Please review these in detail the [Job API](/docs/reference/kubernetes-api/workload-resources/job-v1/).
> We also recommend you running the e2e conformance tests for the Job object to
> verify your implementation.

Reviewer: ```suggestion
Please review these in detail in the [Job API](/docs/reference/kubernetes-api/workload-resources/job-v1/).
We also recommend that you run the e2e conformance tests for the Job object to
verify your implementation.
```

Author: Sorry, I was too slow to handle that. @atiratree do you think these are important enough for follow up PR, or just let it stay as is?

Reviewer: @mimowo  I would prefer an update if you agree that these formulations sound a bit off. I am not saying my version is the best though, so I will leave the final version up to you..

Author: I think both suggestions are slightly better than original form so let me open the PR, we will see if other folks have other proposals there

Author: Follow up: https://github.com/kubernetes/website/pull/45506

  terminating: 3 # three Pods are terminating and have not yet reached the Failed phase
```

### Delegation of managing a Job object to external controller

{{< feature-state for_k8s_version="v1.30" state="alpha" >}}

{{< note >}}
You can only set the `managedBy` field on Jobs if you enable the `JobManagedBy`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
(disabled by default).
{{< /note >}}

This feature allows you to disable the built-in Job controller, for a specific
Job, and delegate reconciliation of the Job to an external controller.

You indicate the controller that reconciles the Job by setting a custom value
for the `spec.managedBy` field - any value

> other than `kubernetes.io/job-controller` - you indicate the controller which
> reconciles the Job. The value of the field is immutable.

Reviewer 2: ```suggestion
other than `kubernetes.io/job-controller`. The value of the field is immutable.
```

Author: done
